### PR TITLE
Update Python chapter

### PR DESCRIPTION
**Pull request by egpbos**
- [x] I followed the [CONTRIBUTING guidelines](../blob/main/CONTRIBUTING.md).

Below, describe what this Pull Request adds:

Was trying to fix some links, but came across prospector references, so I removed those first (one of the broken links was to our prospector config file in the python-template, which no longer exists there) and added mention of Ruff.

~Will fix other links later.~

Fixed other links as well and updated some details on Conda (miniforge now includes mamba, mambaforge is discouraged).

### CODE REVIEW

File: best_practices/language_guides/python.md [hunk 1]
```diff
@@ -137,16 +139,14 @@ For packaging your code, you can either use `pip` or `conda`. Neither of them is
...
+We have long promoted use of [`prospector`](https://github.com/landscapeio/prospector), a tool for running a suite of linters, including, among others [pycodestyle](https://github.com/PyCQA/pycodestyle), [pydocstyle](https://github.com/PyCQA/pydocstyle), [pyflakes](https://pypi.python.org/pypi/pyflakes), [pylint](https://www.pylint.org/), [mccabe](https://github.com/PyCQA/mccabe) and [pyroma](https://github.com/regebro/pyroma).
 
-* [pycodestyle](https://github.com/PyCQA/pycodestyle)
-* [pydocstyle](https://github.com/PyCQA/pydocstyle)
-* [pyflakes](https://pypi.python.org/pypi/pyflakes)
-* [pylint](https://www.pylint.org/)
-* [mccabe](https://github.com/PyCQA/mccabe)
-* [pyroma](https://github.com/regebro/pyroma)
+However, we have [since 2023 been switching](https://github.com/NLeSC/python-template/issues/336) to [Ruff](https://github.com/astral-sh/ruff).
+It is much faster and aims to support most of the functionality that `prospector` does (see the website for the complete function parity overview).
```
Comment (bouweandela) [inline best_practices/language_guides/python.md:146]: A popular alternative to the suite of linters mentioned above is [Ruff](https://github.com/astral-sh/ruff) (see the website for the complete function parity overview).

File: best_practices/language_guides/python.md [hunk 2]
```diff
@@ -137,16 +139,14 @@ For packaging your code, you can either use `pip` or `conda`. Neither of them is
...
 
-* [pycodestyle](https://github.com/PyCQA/pycodestyle)
-* [pydocstyle](https://github.com/PyCQA/pydocstyle)
-* [pyflakes](https://pypi.python.org/pypi/pyflakes)
-* [pylint](https://www.pylint.org/)
-* [mccabe](https://github.com/PyCQA/mccabe)
-* [pyroma](https://github.com/regebro/pyroma)
+However, we have [since 2023 been switching](https://github.com/NLeSC/python-template/issues/336) to [Ruff](https://github.com/astral-sh/ruff).
+It is much faster and aims to support most of the functionality that `prospector` does (see the website for the complete function parity overview). 
+It can be configured in a `pyproject.toml` section.
```
Comment (bouweandela) [inline best_practices/language_guides/python.md:147]: This is probably for all of the abovementioned tools, not just for ruff.

File: best_practices/language_guides/python.md [hunk 3]
```diff
@@ -137,16 +139,14 @@ For packaging your code, you can either use `pip` or `conda`. Neither of them is
 
 The style guide for Python code is [PEP8](http://www.python.org/dev/peps/pep-0008/) and for docstrings it is [PEP257](https://www.python.org/dev/peps/pep-0257/). We highly recommend following these conventions, as they are widely agreed upon to improve readability. To make following them significantly easier, we recommend using a linter.
 
-Many linters exists for Python, [`prospector`](https://github.com/landscapeio/prospector) is a tool for running a suite of linters, it supports, among others:
+Many linters exists for Python.
+We have long promoted use of [`prospector`](https://github.com/landscapeio/prospector), a tool for running a suite of linters, including, among others [pycodestyle](https://github.com/PyCQA/pycodestyle), [pydocstyle](https://github.com/PyCQA/pydocstyle), [pyflakes](https://pypi.python.org/pypi/pyflakes), [pylint](https://www.pylint.org/), [mccabe](https://github.com/PyCQA/mccabe) and [pyroma](https://github.com/regebro/pyroma).
```
Comment (bouweandela) [inline best_practices/language_guides/python.md:143]: Could you rephrase this to avoid 'We have long promoted'? The guide should provide up to date advice, not an historical account of things we promoted in the past. I think prospector is still a decent solution, but maybe not the most popular at the moment. Still, it runs the tools maintained by the https://github.com/PyCQA community, which may be more reliable for some. Though I see that pydocstyle is now deprecated by that community in favour of ruff, so judging by that we should probably mention ruff first and then prospector as an alternative?

File: best_practices/language_guides/python.md [hunk 4]
```diff
@@ -288,7 +288,7 @@ It is good practice to restart the kernel and run the notebook from start to fin
 
 ### Database Interface
 
-* [psycopg](http://initd.org/psycopg/) is an [PostgreSQL](http://www.postgresql.org) adapter
+* [psycopg](https://www.psycopg.org/) is a [PostgreSQL](http://www.postgresql.org) adapter
```
Comment (bouweandela) [inline best_practices/language_guides/python.md:291]: I'm no expert on databases, but maybe we should we recommend https://www.sqlalchemy.org/ or something similarly database backend independent instead of all these?

**Review by bouweandela — APPROVED** [review]
Looks good already! Just some suggestions for further improvement

File: best_practices/language_guides/python.md (hunk 2 above)
Comment (egpbos) [reply best_practices/language_guides/python.md:147]: True, I'll remove the section then (it was a replacement of an existing config sentence, but since it is now the default, no need to mention separately).

File: best_practices/language_guides/python.md (hunk 4 above)
Comment (egpbos) [reply best_practices/language_guides/python.md:291]: Me neither, I don't feel that qualified to say much about this, although all the db connections from Python I personally have seen / used were indeed with sqlalchemy... Maybe ask a contribution from a recent user.

File: best_practices/language_guides/python.md (hunk 3 above)
Comment (egpbos) [reply best_practices/language_guides/python.md:143]: Ok, I changed to something along these lines.

File: best_practices/language_guides/python.md (hunk 1 above)
Comment (egpbos) [reply best_practices/language_guides/python.md:146]: Included this remark in a slightly modified way because of the reordering suggested above.

**Comment by jiskattema** [comment]
hi all! i'm not sure this is the right place, please let me know if not ;)
I'm using a language server whenever i can, and for python the best (only?) is [https://github.com/python-lsp/python-lsp-server](https://github.com/python-lsp/python-lsp-server)
This comes with its own choice for linter/formatter etc. I really recommend using an LSP to catch errors and find docs!

File: best_practices/language_guides/python.md (hunk 4 above)
Comment (egpbos) [reply best_practices/language_guides/python.md:291]: #316

**Comment by egpbos** [comment]
@jiskattema let's discuss in #317 :)

**Comment by egpbos** [comment]
@bouweandela I merged, since you approved before, but if you'd like to still check the changes, see the last commit :)

File: best_practices/language_guides/python.md [hunk 5]
```diff
@@ -140,16 +140,17 @@ For packaging your code, you can either use `pip` or `conda`. Neither of them is
...
+Some of these tools have seen decreasing community support recently, but it is still a good alternative, having been a defining community default for years.
 
 Most of the above tools can be integrated in text editors and IDEs for convenience.
 
 Autoformatting tools like [`yapf`](https://github.com/google/yapf) and [`black`](https://black.readthedocs.io/en/stable/index.html) can automatically format code for optimal readability. `yapf` is configurable to suit your (team's) preferences, whereas `black` enforces the style chosen by the `black` authors. The [`isort`](http://timothycrosley.github.io/isort/) package automatically formats and groups all imports in a standard, readable way.
 
+Ruff can do autoformatting as well and can function as a drop-in replacement of `black` and `isort`.
```
Comment (bouweandela) [inline best_practices/language_guides/python.md:152]: As far as I'm aware, Ruff can only correct small bits of lint, not full autofomatting..

**Comment by bouweandela** [comment]
> I merged, since you approved before, but if you'd like to still check the changes, see the last commit :)

Thanks! Looks great now, except for the addition on auto formatting. I created a new issue for it..
